### PR TITLE
feat: add reconnect action.

### PR DESCRIPTION
**Pull request by iFurySt**
1. when `refreshing the page` or `socket reconnect`: try to reconnect to the already initialized agent. if it doesn't exist, auto init.
2. `reset button`: init agent includes related envs like docker containers.
3. `change settings`: init agent, same as step 2.

https://github.com/OpenDevin/OpenDevin/assets/16201837/e3c05ed7-dc19-45be-bd71-b36136f6053e

### CODE REVIEW

**Review by rbren — APPROVED** [review]

File: frontend/src/components/AgentControlBar.tsx [hunk 1]
```diff
@@ -72,13 +76,24 @@ function ActionButton({
...
   const [desiredState, setDesiredState] = React.useState(AgentTaskState.INIT);
   const [isLoading, setIsLoading] = React.useState(false);
 
+  const handleReset = () => {
+    if (!initialized) return;
+
+    initializeAgent(getSettings());
+    // act = AgentTaskAction.STOP;
+    clearMsgs().then().catch();
+    store.dispatch(clearMessages());
```
Comment (amanape) [inline frontend/src/components/AgentControlBar.tsx:94]: Avoid importing the store directly into components. Always use hooks to interact with the store (`useDispatch` in this case)

File: frontend/src/components/AgentControlBar.tsx [hunk 2]
```diff
@@ -72,13 +76,24 @@ function ActionButton({
...
   const { curTaskState } = useSelector((state: RootState) => state.agent);
   const [desiredState, setDesiredState] = React.useState(AgentTaskState.INIT);
   const [isLoading, setIsLoading] = React.useState(false);
 
+  const handleReset = () => {
+    if (!initialized) return;
+
+    initializeAgent(getSettings());
+    // act = AgentTaskAction.STOP;
+    clearMsgs().then().catch();
```
Comment (amanape) [inline frontend/src/components/AgentControlBar.tsx:93]: Prefer `async...await`. It is a cleaner implementation over an empty `then...catch`

**Review by amanape — CHANGES_REQUESTED** [review]

File: frontend/src/components/AgentControlBar.tsx (hunk 1 above)
Comment (amanape) [reply frontend/src/components/AgentControlBar.tsx:94]: https://redux.js.org/faq/store-setup#can-or-should-i-create-multiple-stores-can-i-import-my-store-directly-and-use-it-in-components-myself

File: frontend/src/components/AgentControlBar.tsx (hunk 2 above)
Comment (iFurySt) [reply frontend/src/components/AgentControlBar.tsx:93]: let me adjust it

**Review by amanape — APPROVED** [review]

**Review by basedanarki — APPROVED** [review]
LGTM +1 strongly typed docstring

**Comment by rbren** [comment]
@iFurySt lgtm! Just need to fix the conflicts

**Comment by iFurySt** [comment]
@rbren okay, i'll handle it tomorrow.

**Comment by neubig** [comment]
Hi @iFurySt , I'm just looking through the backlog and was just wondering what you'd like to do with this PR? Fix the conflicts? Close and possibly do in a different PR? Any way is fine, just checking!

**Comment by iFurySt** [comment]
> Hi @iFurySt , I'm just looking through the backlog and was just wondering what you'd like to do with this PR? Fix the conflicts? Close and possibly do in a different PR? Any way is fine, just checking!

I'll close this PR, it's been behind too long.

**Pull request closed without merging**